tests/future/batch_semaphore.rs

```diff
@@ -0,0 +1,654 @@
+                    check_clock(|i, c| (i == 3) || (c > 0));
+                });
+            },
+            None,
+        );
+    }
+}
+
+#[test]
+fn batch_semaphore_clock_3() {
+    for fairness in [Fairness::StrictlyFair, Fairness::Unfair] {
+        check_dfs(
+            move || {
+                let s = Arc::new(BatchSemaphore::new(0, fairness));
+
+                for i in 1..=2 {
+                    let s2 = s.clone();
+                    thread::spawn(move || {
+                        assert_eq!(me(), i);
+                        s2.release(1);
+                    });
+                }
+
+                thread::spawn(move || {
+                    assert_eq!(me(), 3);
+                    check_clock(|i, c| (c > 0) == (i == 0));
+                    // acquire 1: unblocked once either of the threads finished
+                    s.acquire_blocking(1).unwrap();
+                    // after the acquire, we are causally dependent on exactly one of the two tasks
+                    let clock = current::clock();
+                    assert!((clock[1] > 0 && clock[2] == 0) || (clock[1] == 0 && clock[2] > 0));
+                });
+            },
+            None,
+        );
+    }
+}
+
+#[test]
+fn batch_semaphore_clock_4() {
+    for fairness in [Fairness::StrictlyFair, Fairness::Unfair] {
+        check_dfs(
+            move || {
+                let s = Arc::new(BatchSemaphore::new(1, fairness));
+
+                for tid in 1..=2 {
+                    let other_tid = 2 - tid;
+                    let s2 = s.clone();
+                    thread::spawn(move || {
+                        assert_eq!(me(), tid);
+                        match s2.try_acquire(1) {
+                            Ok(()) => {
+                                // we won the race, no causal dependence on another thread
+                                check_clock(|i, c| (c > 0) == (i == 0 || i == tid));
+                            }
+                            Err(TryAcquireError::NoPermits) => {
+                                // we lost the race, so we causally depend on the other thread
+                                check_clock(|i, c| !(i == 0 || i == other_tid) || (c > 0));
+                            }
+                            Err(TryAcquireError::Closed) => unreachable!(),
+                        }
+                    });
+                }
+            },
+            None,
+        );
+    }
+}
+
+/// Shows a case in which causality tracking in the batch semaphore is
+/// currently imprecise. The test sets up a semaphore with two permits and two
+/// threads, each of which acquires one permit, then releases one permit, then
+/// repeats. Neither thread can be blocked in this situation, and so neither
+/// thread should causally depend on the other, but currently they do.
+#[test]
+#[should_panic(expected = "doesn't satisfy predicate")]
+fn batch_semaphore_clock_imprecise() {
```

Comment:
@jorajeev Here is a test which currently fails but in theory should succeed if our causality tracking was "precise".